File: Week_03/powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        # 采用内部函数，代码更加紧凑
        def quickMul(n: int) -> float:
            # terminator
            if n == 0: return 1.0
            # current logic and drill down
            # 这里必须使用 '//' 运算符，结果才是 int，使用 '/' 运算符结果为 float
            res = quickMul(n // 2)
            # combine results
            return res * res if n % 2 == 0 else res * res * x

        return quickMul(n) if n >= 0 else 1.0 / quickMul(-n)

File: Week_03/powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        long long N = n;

        return N >= 0 ? quickMul(x, N) : 1.0 / quickMul(x, -N);
    }

    double quickMul(double x, long long n) {
        // terminator
        if (n == 0) return 1;
        // current level logic, drill down, combine result
        double res = quickMul(x, n / 2);
        return n & 1 ? res * res * x : res * res;
    }
};
